### ISSUE TITLE
Circular reference results in deadlock

### ISSUE THREAD

**Issue by davidfowl** [dotnet/runtime#36458]
In 3.0 we rewrote the code generation logic in the DI container to make it resillent to stack overflows for deep object graphs. As a result of this change, what would previously result in a stackoverflow exception for lazily resolved circular references now results in ~~an infinite recursion. We're missing a number of stacks maximum.~~ a deadlock.

```C#
using System;
using Microsoft.Extensions.DependencyInjection;

namespace WebApplication315
{
    public class Program
    {
        public static void Main(string[] args)
        {
            var services = new ServiceCollection();
            services.AddSingleton<Foo>();
            var serviceProvider = services.BuildServiceProvider();
            var foo = serviceProvider.GetService<Foo>();
        }
    }

    public class Foo
    {
        public Foo(IServiceProvider serviceProvider)
        {
            serviceProvider.GetService<Foo>();
        }
    }
}
```

**Comment by ericstj** [comment]
Could be similar to https://github.com/dotnet/runtime/issues/35986

**Comment by eerhardt** [comment]
@maryamariyan - was this fixed with your deadlock work in 6.0? Moving to 7.0, in case it wasn't. But please close if this is fixed in the latest.

**Comment by davidfowl** [comment]
No this isn't fixed.

**Comment by springy76** [comment]
I'm on net5 using M.E.DI 6.0.0-preview7 and exactly got that deadlock on StackGuard>WaitOne yesterday :-|

**Comment by EnCey** [comment]
I believe I also just fell in this hole, on net6.0. Took quite some digging around to figure out that a circular reference was the culprit, I was certain we'd get an exception from the DI container if that happened and didn't even look at that when I started my investigation 🙈

**Comment by davidfowl** [comment]
The fix here is easy but adds overhead. We should prototype something. If anybody wants to look at this, we do a cycle check when building up the callsite, this let's use detect constructor visible dependency at startup. This doesn't work for runtime resolved dependencies as the [`CallSiteRuntimeResolver`](https://github.com/dotnet/runtime/blob/911cc41358aa3a99c24fa85dca426fafa3bef555/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs) doesn't detect cycles.

**Comment by lawrence-laz** [comment]
Is there a work around to detect this during run time? 
We are working on a big code base that is slowly moving to dependency containers and I can see this coming back to bite us many times...

**Comment by wvpm** [comment]
Moved from #105900 

### Possible solution
I think the loop is caused by [the lock inside the CallSiteFactory](https://github.com/dotnet/runtime/blob/f120fff4838fbae622123c6578734275d909a3f2/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs#L197). I might be wrong, but it's a good place to look for the bug.

I think changing the code as follows might work:
```cs
            var callsiteLock = _callSiteLocks.GetOrAdd(serviceIdentifier, static _ => new object());
+           callSiteChain.CheckCircularDependency(serviceIdentifier);
            lock (callsiteLock)
            {
-               callSiteChain.CheckCircularDependency(serviceIdentifier);

                ServiceCallSite? callSite = TryCreateExact(serviceIdentifier, callSiteChain) ??
                                           TryCreateOpenGeneric(serviceIdentifier, callSiteChain) ??
                                           TryCreateEnumerable(serviceIdentifier, callSiteChain);

                return callSite;
            }
```

**Comment by wvpm** [comment]
@davidfowl 
I fixed it in https://github.com/wvpm/runtime/commit/f47317a6bd67ca9f7dedda944bf9f0fad1dd821d .
All tests read green on my machine, including the scenario I added in #105900 .

**Comment by davidfowl** [comment]
Is there any reason you didn't do the check in the CallSiteRuntimeResolver?

**Comment by wvpm** [comment]
> Is there any reason you didn't do the check in the CallSiteRuntimeResolver?

I tried that initially but it doesn't work.
You could either store the state in the `CallSiteRuntimeResolver` or pass it through the functions.
Storing it in the `CallSiteRuntimeResolver` caused deadlocks anyway, I think it's due to the static Instance.
Passing it through the functions doesn't work either because the `Resolve` method gets called halfway through, creating new state.

Storing the state in the `ServiceProviderEngineScope ` works perfect.
I should note that I deleted the remote execution tests as I couldn't get them to run and thus haven't tested them.

**Comment by davidfowl** [comment]
Wouldn't it be state on the RuntimeResolverContext?

**Comment by wvpm** [comment]
> Wouldn't it be state on the RuntimeResolverContext?

@davidfowl 
No, because a new context is created for every call to `CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)`. Only the `ServiceProviderEngineScope` is passed throughout the process. That's why I added the state there.

**Comment by wvpm** [comment]
@davidfowl is there anything more I can do to fix this?

The solutions provided in https://github.com/wvpm/runtime/commit/f47317a6bd67ca9f7dedda944bf9f0fad1dd821d + https://github.com/wvpm/runtime/commit/8081e734abb6e5b20e8401b8a514171511340c30 should suffice.
If those changes aren't sufficient, please let me know and I can have a look at it again.
If it's merely a question of planning internally, consider dropping the "help wanted" tag.

**Comment by rosebyte** [comment]
Triage: Seems addressed in #119198

**Comment by rosebyte** [comment]
Exposing DI in ways that external actors can misuse opens the door to a lot of problems, it’s not specific to this particular case. We also seem to agree there’s no legitimate use case for circular references (you pointed that out from the beginning, and I agree). Putting that aside, what remains is that an explicit exception would be better than a deadlock, which we both agree on.

However, the scenario is niche and completely avoidable through functional tests and by designing with a predictable class dependency graph. Fixing it generically would likely add overhead that every application would pay for.

This isn’t a rejection. The previous issue closure wasn’t meant that way either. We’re just keeping a single issue open for this problem. I’ll review the related ones and pick one to represent the rest.

If the current spike finds a solution without perf or memory overhead, we’ll take it. Otherwise, we’ll keep it in mind if we get a chance to redesign DI internals in the future.
